include/materials/material_utility.tcc

```diff
@@ -0,0 +1,230 @@
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  const double j2 = (std::pow((stress(0) - stress(1)), 2) +
+                     std::pow((stress(1) - stress(2)), 2) +
+                     std::pow((stress(0) - stress(2)), 2)) /
+                        6.0 +
+                    std::pow(stress(3), 2) + std::pow(stress(4), 2) +
+                    std::pow(stress(5), 2);
+
+  return j2;
+}
+
+//! Compute J3 invariant
+inline const double mpm::material_utility::compute_j3(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute deviatoric stress
+  Eigen::Matrix<double, 6, 1> dev_stress = compute_deviatoric_stress(stress);
+
+  // Compute J3
+  double j3 = (dev_stress(0) * dev_stress(1) * dev_stress(2)) -
+              (dev_stress(2) * std::pow(dev_stress(3), 2)) +
+              ((2 * dev_stress(3) * dev_stress(4) * dev_stress(5)) -
+               (dev_stress(0) * std::pow(dev_stress(4), 2)) -
+               (dev_stress(1) * std::pow(dev_stress(5), 2)));
+
+  return j3;
+}
+
+//! Compute deviatoric q
+inline const double mpm::material_utility::compute_deviatoric_q(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute J2 from
+  const double j2 = compute_j2(stress);
+
+  const double deviatoric_q = std::sqrt(3 * j2);
+
+  return deviatoric_q;
+}
+
+//! Compute Lode angle
+inline const double mpm::material_utility::compute_lode_angle(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute j2 and j3
+  const double j2 = compute_j2(stress);
+  const double j3 = compute_j3(stress);
+
+  // Compute Lode angle value
+  double lode_angle_val = 0.0;
+  if (abs(j2) > 1.0E-6) {
+    lode_angle_val = (3. * std::sqrt(3.) / 2.) * (j3 / std::pow(j2, 1.5));
+  }
+  if (lode_angle_val > 1.0) lode_angle_val = 1.0;
+  if (lode_angle_val < -1.0) lode_angle_val = -1.0;
+
+  // Compute Lode angle (sin convention)
+  double lode_angle = (1. / 3.) * acos(lode_angle_val);
+  if (lode_angle > M_PI / 3.) lode_angle = M_PI / 3.;
+  if (lode_angle < -0.) lode_angle = 0.;
+
+  return lode_angle;
+}
+
+//! Compute derivative of p in terms of stress sigma
+inline const Eigen::Matrix<double, 6, 1>
+    mpm::material_utility::compute_dp_dsigma(
+        const Eigen::Matrix<double, 6, 1>& stress) {
+
+  Eigen::Matrix<double, 6, 1> dp_dsigma = Eigen::Matrix<double, 6, 1>::Zero();
+  dp_dsigma(0) = 1. / 3.;
+  dp_dsigma(1) = 1. / 3.;
+  dp_dsigma(2) = 1. / 3.;
+
+  return dp_dsigma;
+}
+
+//! Compute derivative of q in terms of stress sigma
+inline const Eigen::Matrix<double, 6, 1>
+    mpm::material_utility::compute_dq_dsigma(
+        const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute q
+  const double deviatoric_q = compute_deviatoric_q(stress);
+
+  // Compute deviatoric stress
+  Eigen::Matrix<double, 6, 1> dev_stress = compute_deviatoric_stress(stress);
+
+  // Compute dq / dsigma
+  Eigen::Matrix<double, 6, 1> dq_dsigma = Eigen::Matrix<double, 6, 1>::Zero();
+  if (abs(deviatoric_q) > 1.E-6) {
```

Comment:
Do you think this tolerance is too large? Do we need to define a same tolerance for all denominator?

Reply:
I don't mind using 1.E-15 actually, but it will blow up more right?

Reply:
We should pass a default tolerance value

Reply:
OK

Reply:
@kks32 isn't better to create `mpm::tolerance()` so that by default all tolerance is the same by default? Or we can utilize `std::numeric_limits<double>::epsilon()`

Reply:
I have used the latter actually `std::numeric_limits<double>::epsilon()`